File: 2025/WishlistApp/backend/src/main/java/com/wishlist/service/WishlistItemService.java
package com.wishlist.service;

import com.wishlist.model.WishlistItem;
import com.wishlist.repository.WishlistItemRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
public class WishlistItemService {

    @Autowired
    private WishlistItemRepository wishlistItemRepository;

    public List<WishlistItem> getAllItems() {
        return wishlistItemRepository.findAll();
    }

    public WishlistItem getItemById(Long id) {
        Optional<WishlistItem> item = wishlistItemRepository.findById(id);
        return item.orElse(null);
    }

    public List<WishlistItem> searchItems(String query) {
        return wishlistItemRepository.searchByNameOrDescription(query);
    }

    public List<WishlistItem> getItemsByCategory(String category) {
        return wishlistItemRepository.findByCategory(category);
    }

    public List<WishlistItem> getItemsByStatus(WishlistItem.Status status) {
        return wishlistItemRepository.findByStatus(status);
    }

    public List<WishlistItem> getItemsByPriority(WishlistItem.Priority priority) {
        return wishlistItemRepository.findByPriority(priority);
    }

    public WishlistItem createItem(WishlistItem item) {
        return wishlistItemRepository.save(item);
    }

    public WishlistItem updateItem(Long id, WishlistItem itemDetails) {
        Optional<WishlistItem> optionalItem = wishlistItemRepository.findById(id);
        if (optionalItem.isPresent()) {
            WishlistItem item = optionalItem.get();
            item.setName(itemDetails.getName());
            item.setDescription(itemDetails.getDescription());
            item.setPrice(itemDetails.getPrice());
            item.setPriority(itemDetails.getPriority());
            item.setStatus(itemDetails.getStatus());
            item.setCategory(itemDetails.getCategory());
            item.setLink(itemDetails.getLink());
            item.setNotes(itemDetails.getNotes());
            return wishlistItemRepository.save(item);
        }
        return null;
    }

    public WishlistItem markAsPurchased(Long id) {
        Optional<WishlistItem> optionalItem = wishlistItemRepository.findById(id);
        if (optionalItem.isPresent()) {
            WishlistItem item = optionalItem.get();
            item.setStatus(WishlistItem.Status.PURCHASED);
            return wishlistItemRepository.save(item);
        }
        return null;
    }

    public boolean deleteItem(Long id) {
        if (wishlistItemRepository.existsById(id)) {
            wishlistItemRepository.deleteById(id);
            return true;
        }
        return false;
    }
}

File: 2025/WishlistApp/frontend/src/app/services/wishlist-item.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { WishlistItem } from '../models/wishlist-item';

@Injectable({
  providedIn: 'root'
})
export class WishlistItemService {
  private apiUrl = 'http://localhost:8095/api/items';

  constructor(private http: HttpClient) {}

  getAllItems(): Observable<WishlistItem[]> {
    return this.http.get<WishlistItem[]>(this.apiUrl);
  }

  getItemById(id: number): Observable<WishlistItem> {
    return this.http.get<WishlistItem>(`${this.apiUrl}/${id}`);
  }

  searchItems(query: string): Observable<WishlistItem[]> {
    const params = new HttpParams().set('query', query);
    return this.http.get<WishlistItem[]>(`${this.apiUrl}/search`, { params });
  }

  getItemsByStatus(status: string): Observable<WishlistItem[]> {
    return this.http.get<WishlistItem[]>(`${this.apiUrl}/status/${status}`);
  }

  getItemsByCategory(category: string): Observable<WishlistItem[]> {
    return this.http.get<WishlistItem[]>(`${this.apiUrl}/category/${category}`);
  }

  getItemsByPriority(priority: string): Observable<WishlistItem[]> {
    return this.http.get<WishlistItem[]>(`${this.apiUrl}/priority/${priority}`);
  }

  createItem(item: WishlistItem): Observable<WishlistItem> {
    return this.http.post<WishlistItem>(this.apiUrl, item);
  }

  updateItem(id: number, item: WishlistItem): Observable<WishlistItem> {
    return this.http.put<WishlistItem>(`${this.apiUrl}/${id}`, item);
  }

  markAsPurchased(id: number): Observable<WishlistItem> {
    return this.http.put<WishlistItem>(`${this.apiUrl}/${id}/purchase`, {});
  }

  deleteItem(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }
}

